barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/scalar_multiplication.cpp

```diff
@@ -220,30 +221,46 @@ void compute_wnaf_states(uint64_t* point_schedule,
         uint64_t* wnaf_table = &point_schedule[(2 * i) * num_initial_points_per_thread];
-        const Fr* thread_scalars = &scalars[i * num_initial_points_per_thread];
         bool* skew_table = &input_skew_table[(2 * i) * num_initial_points_per_thread];
-        uint64_t offset = i * num_points_per_thread;
-
-        for (uint64_t j = 0; j < num_initial_points_per_thread; ++j) {
-            T0 = thread_scalars[j].from_montgomery_form();
-            Fr::split_into_endomorphism_scalars(T0, T0, *(Fr*)&T0.data[2]);
-
-            wnaf::fixed_wnaf_with_counts(&T0.data[0],
-                                         &wnaf_table[(j << 1UL)],
-                                         skew_table[j << 1ULL],
+        // Our offsets for this thread
+        const uint64_t point_offset = i * num_points_per_thread;
+        const uint64_t scalar_offset = i * num_initial_points_per_thread;
+
+        // How many defined scalars are there?
+        const size_t defined_extent = std::min(scalar_offset + num_initial_points_per_thread, scalars.size());
+        const size_t defined_scalars = defined_extent > scalar_offset ? defined_extent - scalar_offset : 0;
+
+        auto wnaf_first_half = [&](const uint64_t* scalar, size_t j) {
+            wnaf::fixed_wnaf_with_counts(scalar,
+                                         &wnaf_table[j * 2],
+                                         skew_table[j * 2],
                                          &thread_round_counts[i][0],
-                                         ((j << 1ULL) + offset) << 32ULL,
+                                         (j * 2 + point_offset) << 32,
                                          num_points,
                                          wnaf_bits);
-            wnaf::fixed_wnaf_with_counts(&T0.data[2],
-                                         &wnaf_table[(j << 1UL) + 1],
-                                         skew_table[(j << 1UL) + 1],
+        };
+        auto wnaf_second_half = [&](const uint64_t* scalar, size_t j) {
+            wnaf::fixed_wnaf_with_counts(scalar,
+                                         &wnaf_table[j * 2 + 1],
+                                         skew_table[j * 2 + 1],
                                          &thread_round_counts[i][0],
-                                         ((j << 1UL) + offset + 1) << 32UL,
+                                         (j * 2 + point_offset + 1) << 32,
                                          num_points,
                                          wnaf_bits);
+        };
+        for (uint64_t j = 0; j < defined_scalars; j++) {
+            Fr T0 = scalars[scalar_offset + j].from_montgomery_form();
+            Fr::split_into_endomorphism_scalars(T0, T0, *(Fr*)&T0.data[2]);
+
+            wnaf_first_half(&T0.data[0], j);
+            wnaf_second_half(&T0.data[2], j);
+        }
+        for (uint64_t j = defined_scalars; j < num_initial_points_per_thread; j++) {
+            // If we are trying to use a non-power-of-2
```

Comment:
This could be simpler, I just took the obviously correct path as I didn't want to spend too long on it

Reply:
Either waway, thanks for clarifying it a bit :)